Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -0,0 +1,222 @@
+            (if Even m then 1 else b ^ 4)
+    else
+      let m := n / 2
+      have h5 : m + 5 < n + 5 := by
+        linarith only [show n = 2 * m + 1
+          by exact (Nat.two_mul_div_two_add_one_of_odd <| Nat.odd_iff_not_even.mpr hn).symm]
+      have h4 : m + 4 < n + 5 := (lt_add_one _).trans h5
+      have h3 : m + 3 < n + 5 := (lt_add_one _).trans h4
+      have h2 : m + 2 < n + 5 := (lt_add_one _).trans h3
+      have h1 : m + 1 < n + 5 := (lt_add_one _).trans h2
+      EllDivSequence'' b c d (m + 2) ^ 2 * EllDivSequence'' b c d (m + 3) *
+          EllDivSequence'' b c d (m + 5) -
+        EllDivSequence'' b c d (m + 1) * EllDivSequence'' b c d (m + 3) *
+            EllDivSequence'' b c d (m + 4) ^ 2
+
+variable (b c d : R)
+
+/-- The canonical example of an elliptic divisibility sequence `h : ℕ → R`,
+with initial values `h(2) = b`, `h(3) = c`, and `h(4) = d`.
+
+This is defined in terms of a truncated sequence whose even terms differ by a factor of `b`. -/
+def EllDivSequence' (n : ℕ) : R :=
+  EllDivSequence'' b c d n * if Even n then b else 1
+
+@[simp]
+lemma EllDivSequence'_zero : EllDivSequence' b c d 0 = 0 := by
+  rw [EllDivSequence', EllDivSequence'', zero_mul]
+
+@[simp]
+lemma EllDivSequence'_one : EllDivSequence' b c d 1 = 1 := by
+  rw [EllDivSequence', EllDivSequence'', one_mul, if_neg Nat.not_even_one]
+
+@[simp]
+lemma EllDivSequence'_two : EllDivSequence' b c d 2 = b := by
+  rw [EllDivSequence', EllDivSequence'', one_mul, if_pos even_two]
+
+@[simp]
+lemma EllDivSequence'_three : EllDivSequence' b c d 3 = c := by
+  rw [EllDivSequence', EllDivSequence'', if_neg <| by decide, mul_one]
+
+@[simp]
+lemma EllDivSequence'_four : EllDivSequence' b c d 4 = d * b := by
+  rw [EllDivSequence', EllDivSequence'', if_pos <| by decide]
+
+@[simp]
+lemma EllDivSequence'_odd (m : ℕ) : EllDivSequence' b c d (2 * (m + 2) + 1) =
+    EllDivSequence' b c d (m + 4) * EllDivSequence' b c d (m + 2) ^ 3 -
+      EllDivSequence' b c d (m + 1) * EllDivSequence' b c d (m + 3) ^ 3 := by
+  rw [EllDivSequence', if_neg <| fun h => Nat.even_add_one.mp h <| even_two_mul _,
+    show 2 * (m + 2) + 1 = 2 * m + 5 by rfl, EllDivSequence'', dif_pos <| even_two_mul m]
+  simp only [EllDivSequence', Nat.mul_div_right _ zero_lt_two]
+  by_cases hm : Even m
+  · have hm1 : ¬Even (m + 1) := fun h => Nat.even_add_one.mp h hm
+    have hm2 : Even (m + 2) := Nat.even_add_one.mpr hm1
+    have hm3 : ¬Even (m + 3) := fun h => Nat.even_add_one.mp h hm2
+    have hm4 : Even (m + 4) := Nat.even_add_one.mpr hm3
+    rw [if_pos hm, if_pos hm, if_pos hm4, if_pos hm2, if_neg hm1, if_neg hm3]
+    ring1
+  · have hm1 : Even (m + 1) := Nat.even_add_one.mpr hm
+    have hm2 : ¬Even (m + 2) := fun h => Nat.even_add_one.mp h hm1
+    have hm3 : Even (m + 3) := Nat.even_add_one.mpr hm2
+    have hm4 : ¬Even (m + 4) := fun h => Nat.even_add_one.mp h hm3
+    rw [if_neg hm, if_neg hm, if_neg hm4, if_neg hm2, if_pos hm1, if_pos hm3]
+    ring1
+
+@[simp]
```

Comment:
why have this as a simp lemma? the RHS doesn't look simpler to me

Reply:
Good point - I will remove this. I think the one for `odd` is similar so I'll remove that as well.